### PR TITLE
CATROID-1105 Scoped storage-external storage files

### PR DESCRIPTION
**Pull request by csum1**
To support scoped storage the system file picker is used for:
- import/export projects
- read/write variable to file bricks
- write embroidery to file

Depending on API version a different storage mechanism is used
- above and equal API29 uses the system file picker activity
- under API29 uses legacy external storage with the FilePickerActivity

The compiled sdk version and target sdk version are set to 30
The flag to request legacy external storage is set to false

Ticket: [CATROID-1105](https://jira.catrob.at/browse/CATROID-1105)

Also solves BUG [CATROID-1284](https://jira.catrob.at/browse/CATROID-1284)

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [x] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/ProjectOptionsFragment.kt [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+    }
+
+    private fun setupNameInputLayout() {
+        nameInputLayout = optionsView?.findViewById(R.id.project_options_name_layout)
+        nameInputLayout?.apply {
+            val editText = editText
+            if (editText != null) {
+                editText.setText(project?.name)
+                editText.addTextChangedListener(object : NewProjectNameTextWatcher<Nameable>() {
+                    override fun afterTextChanged(s: Editable) {
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/ProjectOptionsFragment.kt:118]: rename variable

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/ProjectOptionsFragment.kt [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+                        nameInputLayout?.error = if (s.toString() != project?.name) {
+                            val error = validateInput(s.toString(), getContext())
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/ProjectOptionsFragment.kt:120]: context

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/ProjectOptionsFragment.kt [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+        if (tags?.size == 1 && tags[0].isEmpty()) {
+            tagsLayout?.visibility = View.GONE
+            return
+        }
+        tagsLayout?.visibility = View.VISIBLE
+        if (tags != null) {
+            for (tag in tags) {
+                val context = context
+                if (context != null) {
+                    val chip = Chip(getContext())
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/ProjectOptionsFragment.kt:161]: context

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/ProjectOptionsFragment.kt [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+            notesAndCreditsInputLayout?.editText?.text.toString().trim { it <= ' ' }
+        if (project?.notesAndCredits == null || project?.notesAndCredits != notesAndCredits) {
+            project?.notesAndCredits = notesAndCredits
+            if (!XstreamSerializer.getInstance().saveProject(project)) {
+                ToastUtil.showError(activity, R.string.error_set_notes_and_credits)
+            }
+        }
+    }
+
+    @Suppress("UNUSED_PARAMETER")
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/ProjectOptionsFragment.kt:295]: delete

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/ProjectOptionsFragment.kt [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+    fun projectUpload(view: View?) {
+        val currentProject = ProjectManager.getInstance().currentProject
+        context?.apply {
+            val projectSaver = ProjectSaver(currentProject, this)
+            projectSaver.saveProjectAsync({ success: Boolean -> onSaveProjectComplete(success) })
+            Utils.setLastUsedProjectName(this, currentProject.name)
+        }
+    }
+
+    @Suppress("UNUSED_PARAMETER")
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/ui/fragment/ProjectOptionsFragment.kt:305]: delete

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ProjectListFragment.kt [hunk 6]
```diff
@@ -0,0 +1,555 @@
...
+        try {
+            val project = XstreamSerializer.getInstance().loadProject(item.directory, activity)
+            ProjectManager.getInstance().currentProject = project
+            requireActivity().supportFragmentManager.beginTransaction()
+                .replace(
+                    R.id.fragment_container, ProjectOptionsFragment(), ProjectOptionsFragment.TAG
+                )
+                .addToBackStack(ProjectOptionsFragment.TAG)
+                .commit()
+        } catch (e: IOException) {
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ProjectListFragment.kt:500]: exception

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ProjectListFragment.kt [hunk 7]
```diff
@@ -0,0 +1,555 @@
...
+        fun getLocalProjectList(items: MutableList<ProjectData>) {
+            for (projectDir in FlavoredConstants.DEFAULT_ROOT_DIRECTORY.listFiles()) {
+                val xmlFile = File(projectDir, Constants.CODE_XML_FILE_NAME)
+                if (!xmlFile.exists()) {
+                    continue
+                }
+                val metaDataParser = ProjectMetaDataParser(xmlFile)
+                try {
+                    items.add(metaDataParser.projectMetaData)
+                } catch (e: IOException) {
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ProjectListFragment.kt:549]: exception

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ProjectListFragment.kt [hunk 8]
```diff
@@ -0,0 +1,555 @@
...
+                    Log.e(TAG, "Well, that's awkward.", e)
```
Comment (juliajulie95) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/fragment/ProjectListFragment.kt:550]: could not parse local project

**Review by juliajulie95 — CHANGES_REQUESTED** [review]

File: catroid/src/main/java/org/catrobat/catroid/ui/fragment/ProjectOptionsFragment.kt (hunk 2 above)
Comment (csum1) [reply catroid/src/main/java/org/catrobat/catroid/ui/fragment/ProjectOptionsFragment.kt:120]: an dieser Stelle ist "context" zweideutig, mit getContext() ist es explizit der Context vom Fragment. "context" würde sich auf InputWatcher.context beziehen. Außerhalb von diesem Scope ist "context" dann wieder der Context vom Fragment. Würde es sicherheitshalber so lassen.

Comment (juliajulie95) [reply catroid/src/main/java/org/catrobat/catroid/ui/fragment/ProjectOptionsFragment.kt]: Dann bitte wieder zuurück ändern, jetzt steht "context" drin

**Review by juliajulie95 — APPROVED** [review]

**Pull request merged**